src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs

```diff
@@ -57,15 +57,19 @@ public class ChainLevelHelper : IChainLevelHelper
     {
         long? startingPoint = GetStartingPoint();
         if (startingPoint == null)
+        {
+            if (_logger.IsTrace)
+                _logger.Trace($"ChainLevelHelper.GetNextHeaders - starting point is null");
             return null;
-
-        long destinationPoint = GetDestinationPoint();
+        }
 
         List<BlockHeader> headers = new(maxCount);
         int i = 0;
 
-        long upperLimit = (destinationPoint - blocksToSkip); // blocksToSkip is used for FastSync boundary
-        while (i < maxCount && (startingPoint <= upperLimit))
+        // blocksToSkip is used for FastSync boundary.
+        long upperLimit = (_blockCacheService.ProcessDestination?.Number ?? _blockTree.BestSuggestedBeaconHeader?.Number ?? 0) - blocksToSkip;
```

Comment:
I don't know why we need upper boundary at all

Reply:
Its for FastSync boundary.

Reply:
I think we should remove this upperLimit, but we can discuss it on Monday. It is better than something we have on master, so merging. However, it looks like our sync hive tests are not working.